test/plugins/cookieauth.js

```diff
@@ -0,0 +1,984 @@
+          .post('/_session', {name: ME, password: PASSWORD})
+          .times(3)
+          .reply(500, {error: 'internal_server_error', reason: 'Internal Server Error'});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        assert.equal(resp.statusCode, 500);
+        assert.ok(data.indexOf('"error":"internal_server_error"') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('returns 200 response on authentication retry success', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .times(2)
+          .reply(500, {error: 'internal_server_error', reason: 'Internal Server Error'})
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER)
+          .get(DBNAME)
+          .reply(200, {doc_count: 0});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        assert.equal(resp.request.headers.cookie, MOCK_COOKIE);
+        assert.equal(resp.statusCode, 200);
+        assert.ok(data.indexOf('"doc_count":0') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('renews authentication token on 401 response', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER)
+          .get(DBNAME)
+          .reply(401, {error: 'unauthorized', reason: 'Unauthorized'})
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER_2)
+          .get(DBNAME)
+          .reply(200, {doc_count: 0});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        assert.equal(resp.request.headers.cookie, MOCK_COOKIE_2);
+        assert.equal(resp.statusCode, 200);
+        assert.ok(data.indexOf('"doc_count":0') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('renews authentication token on 403 response', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER)
+          .get(DBNAME)
+          .reply(403, {error: 'forbidden', reason: 'Forbidden'})
```

Comment:
We should **not** try a `_session` request in the case of a 403 with reason `forbidden` we should only try to request a new session in the case of a 403 with reason `credentials_expired`. We should probably have tests to cover both these scenarios instead of just the new session case.

Reply:
As discussed, the data stream isn't currently accessible to the plugin hook so we therefore can't analyse the response body.

The problem is we must choose to retry a request before the data starts flowing. If we don't then we'll be unable to pipe to the client stream as data will have already been emitted. I'll raise an issue to capture this, although I'm not sure how to fix this at the moment.